### PR TITLE
feat(cache):  Explicitly specifying which attributes affect the cache key

### PR DESCRIPTION
**Pull request by Ark-kun**
Fixes https://github.com/kubeflow/pipelines/issues/4038
Fixes https://github.com/kubeflow/pipelines/issues/3972

### CODE REVIEW

File: backend/src/cache/server/mutation_test.go [hunk 1]
```diff
@@ -135,9 +135,9 @@ func TestMutatePodIfCached(t *testing.T) {
 
 func TestMutatePodIfCachedWithCacheEntryExist(t *testing.T) {
 	executionCache := &model.ExecutionCache{
-		ExecutionCacheKey: "f98b62e4625b9f96bac478ac72d88181a37e4f1d6bfd3bd5f53e29286b2ca034",
+		ExecutionCacheKey: "f5fe913be7a4516ebfe1b5de29bcb35edd12ecc776b2f33f10ca19709ea3b2f0",
 		ExecutionOutput:   "testOutput",
-		ExecutionTemplate: `{"name": "test_template"}`,
+		ExecutionTemplate: `{"container":{"command":["echo", "Hello"],"image":"python:3.7"}}`,
```
Comment (jingzhang36) [inline backend/src/cache/server/mutation_test.go:140]: Could you also have a quick test to show that adding name doesn't change the signature?

Comment (Ark-kun) [reply backend/src/cache/server/mutation_test.go:140]: Good idea. I'll add that.

File: backend/src/cache/server/mutation.go [hunk 2]
```diff
@@ -178,14 +191,23 @@ func generateCacheKeyFromTemplate(template string) (string, error) {
...
 	}
 
-	// template[archiveLocation] needs to be removed when calculating cache key.
-	// Because archiveLocation.key is different in every single run.
-	_, exists := templateMap[ArchiveLocationKey]
-	if exists {
-		log.Println("ArchiveLocation exists in template.")
-		delete(templateMap, ArchiveLocationKey)
+	// Selectively copying parts of the template that should affect the cache
+	templateSkeleton := map[string]interface{}{
```
Comment (Bobgy) [inline backend/src/cache/server/mutation.go:195]: nit: shall we initiate this as a readonly constant value?

File: backend/src/cache/server/mutation.go [hunk 3]
```diff
@@ -170,6 +169,20 @@ func MutatePodIfCached(req *v1beta1.AdmissionRequest, clientMgr ClientManagerInt
 	return patches, nil
 }
 
+func intersectStructureWithSkeleton(src map[string]interface{}, skeleton map[string]interface{}) map[string]interface{} {
```
Comment (Bobgy) [inline backend/src/cache/server/mutation.go:172]: interesting util function!
Would you mind adding a brief comment above about its usage?

File: backend/src/cache/server/mutation.go (hunk 2 above)
Comment (Ark-kun) [reply backend/src/cache/server/mutation.go:195]: Hmm. Are you sure Go support constant maps?

File: backend/src/cache/server/mutation.go (hunk 3 above)
Comment (Ark-kun) [reply backend/src/cache/server/mutation.go:172]: Done.

File: backend/src/cache/server/mutation.go (hunk 2 above)
Comment (Bobgy) [reply backend/src/cache/server/mutation.go:197]: I meant, looks like we are not going to modify it, so can we just declare it once as a global variable?

**Comment by Ark-kun** [comment]
I think I've resolved all review comments and all tests pass.

**Comment by Ark-kun** [comment]
@Bobgy Can you please re-LGTM?

**Comment by Bobgy** [comment]
I commented on https://github.com/kubeflow/pipelines/pull/4076/files#r468250208,

but it's not a blocking issue
/lgtm

**Pull request merged**